[Feature] support for ignoring modules

We are using the following function in order to keep certain modules (generally platform specific dependancies like sqlite3) from being bundled into our webpack bundle which will cause any module in the array to be manually required instead:
```
  externals: [
    (function () {
      var IGNORES = [
        'buffer',
        'fs',
        'electron',
        'sqlite3',
        'JSONStream',
        'request',
        'zlib',
        'bindings',
        'leveldown',
        'pouchdb',
        'canvas'
      ];
      return function (context, request, callback) {
        if (IGNORES.indexOf(request) >= 0) {
          return callback(null, "require('" + request + "')");
        }
        return callback();
      };
    })()
  ]
```

When loading a worker that uses any of those modules directly or uses a module that uses them, webpack attempts to trace back these require statements instead of ignoring them as the above config dictates leading to build errors like:
```
    ERROR in ./~/sqlite3/~/node-pre-gyp/lib/info.js
    Module not found: Error: Can't resolve 'aws-sdk' in '/Users/symbyte/projects/my-project/node_modules/sqlite3/node_modules/node-pre-gyp/lib'
     @ ./~/sqlite3/~/node-pre-gyp/lib/info.js 14:14-32
     @ ./~/sqlite3/~/node-pre-gyp/lib ^\.\/.*$
     @ ./~/sqlite3/~/node-pre-gyp/lib/node-pre-gyp.js
     @ ./~/sqlite3/lib/sqlite3.js
     @ ./~/ts-loader?{"useTranspileModule":true}!./~/tslint-loader?{"configFile":"/Users/symbyte/projects/my-project/tslint.json"}!./~/tslint-loader!./src/app/data/workers/geojson-io.worker.ts
```
Is there some way to get `worker-loader` to acknowledge this part of my webpack config, have I done something incorrectly, or is this a bug?

--- a/src/index.js
+++ b/src/index.js
@@ -4,11 +4,16 @@ import validateOptions from 'schema-utils';
 import NodeTargetPlugin from 'webpack/lib/node/NodeTargetPlugin';
 import SingleEntryPlugin from 'webpack/lib/SingleEntryPlugin';
 import WebWorkerTemplatePlugin from 'webpack/lib/webworker/WebWorkerTemplatePlugin';
+import ExternalsPlugin from 'webpack/lib/ExternalsPlugin';
 
 import schema from './options.json';
 import supportWebpack5 from './supportWebpack5';
 import supportWebpack4 from './supportWebpack4';
-import { getDefaultFilename, getDefaultChunkFilename } from './utils';
+import {
+  getDefaultFilename,
+  getDefaultChunkFilename,
+  getExternalsType,
+} from './utils';
 
 let FetchCompileWasmPlugin;
 let FetchCompileAsyncWasmPlugin;
@@ -56,7 +61,11 @@ export function pitch(request) {
     ? options.chunkFilename
     : getDefaultChunkFilename(compilerOptions.output.chunkFilename);
 
-  worker.options = { filename, chunkFilename, globalObject: 'self' };
+  worker.options = {
+    filename,
+    chunkFilename,
+    globalObject: 'self',
+  };
 
   worker.compiler = this._compilation.createChildCompiler(
     'worker',
@@ -77,6 +86,13 @@ export function pitch(request) {
 
   if (FetchCompileAsyncWasmPlugin) {
     new FetchCompileAsyncWasmPlugin().apply(worker.compiler);
+  }
+
+  if (compilerOptions.externals) {
+    new ExternalsPlugin(
+      getExternalsType(compilerOptions),
+      compilerOptions.externals
+    ).apply(worker.compiler);
   }
 
   new SingleEntryPlugin(this.context, `!!${request}`, 'main').apply(

--- a/src/utils.js
+++ b/src/utils.js
@@ -10,6 +10,28 @@ function getDefaultChunkFilename(chunkFilename) {
   return chunkFilename.replace(/\.([a-z]+)(\?.+)?$/i, '.worker.$1$2');
 }
 
+function getExternalsType(compilerOptions) {
+  // For webpack@4
+  if (compilerOptions.output.libraryTarget) {
+    return compilerOptions.output.libraryTarget;
+  }
+
+  // For webpack@5
+  if (compilerOptions.externalsType) {
+    return compilerOptions.externalsType;
+  }
+
+  if (compilerOptions.output.library) {
+    return compilerOptions.output.library.type;
+  }
+
+  if (compilerOptions.output.module) {
+    return 'module';
+  }
+
+  return 'var';
+}
+
 function getWorker(file, content, options) {
   const publicPath =
     typeof options.publicPath === 'undefined'
@@ -48,5 +70,9 @@ function getWorker(file, content, options) {
   })`;
 }
 
-// eslint-disable-next-line import/prefer-default-export
-export { getDefaultFilename, getDefaultChunkFilename, getWorker };
+export {
+  getDefaultFilename,
+  getDefaultChunkFilename,
+  getExternalsType,
+  getWorker,
+};

--- a/test/fixtures/external/entry.js
+++ b/test/fixtures/external/entry.js
@@ -1,0 +1,22 @@
+import Worker from './worker.js';
+
+const worker = new Worker();
+
+let result;
+
+worker.onmessage = function (event) {
+  if (!result) {
+    result = document.createElement("div");
+    result.setAttribute('id', 'result');
+
+    document.body.append(result);
+  }
+
+  result.innerText = JSON.stringify(event.data)
+};
+
+const button = document.getElementById('button');
+
+button.addEventListener('click', () => {
+  worker.postMessage({ postMessage: true })
+});

--- a/test/fixtures/external/index.html
+++ b/test/fixtures/external/index.html
@@ -1,0 +1,13 @@
+<!doctype html>
+<html>
+<head>
+  <meta charset="utf-8">
+  <title>Webpack App</title>
+  <meta name="viewport" content="width=device-width,initial-scale=1">
+</head>
+<body>
+
+<button id="button">Run Action</button>
+
+</body>
+</html>

--- a/test/fixtures/external/worker.js
+++ b/test/fixtures/external/worker.js
@@ -1,0 +1,9 @@
+import foo from 'my-custom-module';
+
+onmessage = function(event) {
+  const workerResult = event.data;
+
+  workerResult.onmessage = true;
+
+  postMessage(workerResult);
+};

--- a/test/loader.test.js
+++ b/test/loader.test.js
@@ -59,4 +59,25 @@ describe('worker-loader', () => {
     expect(getWarnings(stats)).toMatchSnapshot('warnings');
     expect(getErrors(stats)).toMatchSnapshot('errors');
   });
+
+  it('should work with "externals"', async () => {
+    const compiler = getCompiler(
+      './external/entry.js',
+      {},
+      {
+        externals: {
+          'my-custom-module': 'navigator',
+        },
+      }
+    );
+    const stats = await compile(compiler);
+    const result = await getResultFromBrowser(stats);
+
+    expect(getModuleSource('./external/worker.js', stats)).toMatchSnapshot(
+      'module'
+    );
+    expect(result).toMatchSnapshot('result');
+    expect(getWarnings(stats)).toMatchSnapshot('warnings');
+    expect(getErrors(stats)).toMatchSnapshot('errors');
+  });
 });

